CMakeLists.txt

```diff
@@ -31,6 +31,15 @@ project(Charm++ LANGUAGES CXX C ASM VERSION 7.0.0)
 find_package(Threads)
 find_package(OpenMP) # Do this before Fortran, in case we don't have a Fortran compiler
 
+set(CMK_BUILD_MPI 0)
+set(CMK_BUILD_ON_MPI 0)
+set(CMK_CXX_MPI_BINDINGS 1) # Always true, this just checked for a conflict with old mpich versions.
+if(${NETWORK} STREQUAL "mpi")
+  set(CMK_BUILD_MPI 1)
+  set(CMK_BUILD_ON_MPI 1)
+  find_package(MPI REQUIRED)
+endif()
```

Comment:
Could you mention why you moved the MPI detection to before the Fortran detection?

Reply:
If `find_package(MPI)` is called downstream where it originally was _and_ if a fortran compiler is really not available on the system, cmake's `FindMPI.cmake` fails with trying to invoke the fortran compiler as
```
CMake Error: CMAKE_Fortran_COMPILER not set, after EnableLanguage
CMake Error at /usr/share/cmake-3.18/Modules/FindMPI.cmake:1213 (try_compile):
  Failed to configure test project build system.        
Call Stack (most recent call first):                    
  /usr/share/cmake-3.18/Modules/FindMPI.cmake:1240 (_MPI_try_staged_settings)
  /usr/share/cmake-3.18/Modules/FindMPI.cmake:1427 (_MPI_check_lang_works)
  CMakeLists.txt:425 (find_package)
```

When I just reproduced this (to remind myself that this was indeed the case), using
```sh
cmake -DTARGET=LIBS -DNETWORK=mpi -DSMP=off -DCMAKE_C_COMPILER=mpicc -DCMAKE_CXX_COMPILER=mpicxx -DCMAKE_Fortran_COMPILER=mpif90 -DENABLE_FORTRAN=off <charm-src>
```
I found that the problem fortran is enabled upstream is because of this code:
```cmake
# No Fortran support with MSVC
if(NOT CMK_COMPILER STREQUAL "msvc")
  enable_language(Fortran OPTIONAL)
endif()
```
which turns on fortran support. Even though this is optional, FindMPI.cmake now tries to invoke the fortran compiler because on linux I have no msvc and fortran as a language is now enabled in cmake for the project. I see in `FindMPI.cmake` that it will test for all languages of (C CXX Fortran) that are enabled.

I'm not sure about the intent but the above "msvc" code appears to be a bug for me: just because the CMK_COMPILER is not msvc, I don't think that that should *add* (optional) fortran support. Perhaps something like the following would be better:
```cmake
option(ENABLE_FORTRAN "Enable Fortran" ON)
if(ENABLE_FORTRAN AND CMAKE_Fortran_COMPILER)

   # No Fortran support with MSVC
   if(NOT CMK_COMPILER STREQUAL "msvc")
     enable_language(Fortran OPTIONAL)
   endif()

  include(FortranCInterface)
  FortranCInterface_VERIFY()
  ...
else()
  message(STATUS "Charm++: No Fortran support.")
endif()
```
or maybe even
```cmake
if(ENABLE_FORTRAN AND CMAKE_Fortran_COMPILER AND NOT CMK_COMPILER STREQUAL "msvc")
  enable_language(Fortran OPTIONAL)
  include(FortranCInterface)
  FortranCInterface_VERIFY()
  ...
else()
  message(STATUS "Charm++: No Fortran support.")
endif()
```

While I have been using this patch with`-DENABLE_FORTRAN=off`, I have now also tested the above with enabling fortran (and with a fortran compiler present on the system) as
```sh
cmake -DTARGET=LIBS -DNETWORK=mpi -DSMP=off -DCMAKE_C_COMPILER=mpicc -DCMAKE_CXX_COMPILER=mpicxx -DCMAKE_Fortran_COMPILER=mpif90 -DENABLE_FORTRAN=on <charm-src>
```
but that (still) fails during make:
```
/home/jbakosi/code/charm-jbakosi/b/bin/charmc: line 232: -I/home/jbakosi/code/charm-jbakosi/b/include: No such file or directory
Fatal Error by charmc in directory /home/jbakosi/code/charm-jbakosi/b
   Command -I/home/jbakosi/code/charm-jbakosi/b/include -DAMPI_ERRHANDLER_RETURN=0 -DNDEBUG -O3 -O3 -O2 -c /home/jbakosi/code/charm-jbakosi/b/pup_f.f90 -o CMakeFiles/conv-utilf.dir/pup_f.f90.o returned error code 127
charmc exiting...
```
This seems to be some other problem, maybe only with mpi and/or clang?